music_assistant/server/providers/plex/__init__.py

```diff
@@ -0,0 +1,467 @@
+        elif plex_media.type == "album":
+            return await self._parse_album(plex_media)
+        elif plex_media.type == "track":
+            return await self._parse_track(plex_media)
+        elif plex_media.type == "playlist":
+            return await self._parse_playlist(plex_media)
+        return None
+
+    async def _search_track(self, search_query, limit) -> list[PlexTrack]:
+        return await self._run_async(
+            self._plex_library.searchTracks, title=search_query, limit=limit
+        )
+
+    async def _search_album(self, search_query, limit) -> list[PlexAlbum]:
+        return await self._run_async(
+            self._plex_library.searchAlbums, title=search_query, limit=limit
+        )
+
+    async def _search_artist(self, search_query, limit) -> list[PlexArtist]:
+        return await self._run_async(
+            self._plex_library.searchArtists, title=search_query, limit=limit
+        )
+
+    async def _search_playlist(self, search_query, limit) -> list[PlexPlaylist]:
+        return await self._run_async(self._plex_library.playlists, title=search_query, limit=limit)
+
+    async def _search_track_advanced(self, limit, **kwargs) -> list[PlexTrack]:
+        return await self._run_async(self._plex_library.searchTracks, filters=kwargs, limit=limit)
+
+    async def _search_album_advanced(self, limit, **kwargs) -> list[PlexAlbum]:
+        return await self._run_async(self._plex_library.searchAlbums, filters=kwargs, limit=limit)
+
+    async def _search_artist_advanced(self, limit, **kwargs) -> list[PlexArtist]:
+        return await self._run_async(self._plex_library.searchArtists, filters=kwargs, limit=limit)
+
+    async def _search_playlist_advanced(self, limit, **kwargs) -> list[PlexPlaylist]:
+        return await self._run_async(self._plex_library.playlists, filters=kwargs, limit=limit)
+
+    async def _search_and_parse(self, search_coro: Coroutine, parse_coro: Callable):
+        tasks = []
+        async with TaskGroup() as tg:
+            for item in await search_coro:
+                tasks.append(tg.create_task(parse_coro(item)))
+
+        results = []
+        for task in tasks:
+            results.append(task.result())
+
+        return results
+
+    async def _parse_album(self, plex_album: PlexAlbum) -> Album:
+        """Parse a Plex Album response to an Album model object."""
+        album_id = plex_album.key
+        album = Album(
+            item_id=album_id,
+            name=plex_album.title,
+            provider=self.domain,
+        )
+        if plex_album.year:
+            album.year = plex_album.year
+        if thumb := plex_album.firstAttr("thumb", "parentThumb", "grandparentThumb"):
+            album.metadata.images = [MediaItemImage(ImageType.THUMB, thumb, self.instance_id)]
+        if plex_album.summary:
+            album.metadata.description = plex_album.summary
+
+        album.artist = await self._get_data_cached(plex_album.parentKey, Artist)
```

Comment:
please do not do any caching in your module, this is all handled in the outer controllers itself. You should just return the most recent data when requested. Or did you mean to implement a very short cache just to return recurring items such as an album artist in the listing ? How much data do you have here from an artist within album level ? Because we do support returning a simple ItemMapping object here to just return a reference to another nested object.

Reply:
I do have the key of the artist here. The key is unique on a Plex server and I am also always using the key as id when I create a MA Artist object. But I don't have access to the name of the artist or so here. Would that work or would we have to get the artist from the DB first to generate a mapping?

Reply:
I will do some research, but at least I did not find anything directly in the libraries documentation

Reply:
Oh damned - in German we say, sometimes you don't see the forest because of all the trees. There is a parentTitle in addition to the parentKey and its the same for all the stuff in tracks. What shall I set for sort_name, uri and version? I just found examples in the code where the mapping was created from an existing MediaItem

Reply:
Cool! That saves a lot of performance. 
uri is auto generated if you leave it blank from the provider and item_id

Copy the __post_init__ from MediaItem to ItemMapping and you are all set

Reply:
https://github.com/music-assistant/server/blob/main/music_assistant/common/models/media_items.py#L176

Reply:
The version you can leave out if you don't have it, that's just for textual reasons. When a full object is retrieved, it fetches all info and only then is any matching performed.

Reply:
Okay, I think I have got it :-)
Ps. Do I have to set my domain or my instance_id as provider? I am just a bit confused because I just realized that the key is only on a server unique - So if someone configures multiple servers (and not just multiple libraries on the same server, like I do), I can't guarantee that the combination of the domain "plex" and the key is unique. It would just be unique in the plex instance

Reply:
In your case set the instance_id. It accepts both but setting the domain is more suitable for multiple instances of e.g. spotify where an item exists on all accounts. For filebased providers and plex, set it to the instance id.